### PR TITLE
Get upstream dependencies for a package

### LINKED ISSUE #1: Get a list of upstream dependencies
Replicate the feature from [JuliaHub](https://juliahub.com/ui/Packages/AWS/SpCmB/1.29.1?t=2) of getting upstream dependents on your package.

### PR DESCRIPTION
**Pull request by mattBrzezinski**
# Overview
Resolves: #1 

This package is meant to replicate some of the closed-source work of [JuliaHub](https://juliahub.com/), however this fleshes itself out to also include private registries which JuliaHub does not include.

In this PR I'm adding in functionality to get upstream dependents of a certain package. There are two limitations with this however,

1) If two registries have the same name for a package it will retrieve information from both. Example,

```julia
using PkgDeps

find_upstream_dependencies("Foobar")
```

`General` registry has a package call `Foobar`, you also have a registry `Private` which has the same package name. Now you want to find all packages which are using `Foobar`, you will get:

- All packages in `General` which use `Foobar` (this will be the public version)
- All packages in `Private` which use `Foobar` (presumably this is your private version)

You can work around this by specifying which registry to look into:

```julia
using PkgDeps

private_rego = reachable_registries("Private")
find_upstream_dependencies("Foobar"; registries=private_rego)
```

2) Getting dependents will *only* retrieve ones which currently use the package.

For instance if `MyPackage` used `Foobar` in versions `0.1-0.2`, but `MyPackage` is now on `0.3`, trying to find dependents of `Foobar` will not return `MyPackage`.

Thank you to [Pkg.jl](https://github.com/JuliaLang/Pkg.jl/) devs for making this pretty simple to do, maybe when this package fleshes itself out some more we can introduce this functionality into there.

### CODE REVIEW

File: src/PkgDeps.jl [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    upstream_dependencies = String[]
+
+    for rego in registries
+        for (pkg, v) in rego.pkgs
+            path = joinpath(v.registry_path, v.path, "Deps.toml")
+
+            if isfile(path)
+                file_contents = parsefile(path)
+
+                dep_keys = [k for k in collect(keys(file_contents)) if endswith(k, "-0") || k == "0"]
```
Comment (fchorney) [inline src/PkgDeps.jl:118]: what is the significance of the headers that are `[0]` or end in `[blah-0]`? Just curious what that is supposed to denote

File: test/runtests.jl [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using PkgDeps
+using Test# Create a tmp directory
```
Comment (fchorney) [inline test/runtests.jl]: should there be an extra line here? 
```suggestion
using Test
# Create a tmp directory
```

File: src/PkgDeps.jl (hunk 1 above)
Comment (mattBrzezinski) [reply src/PkgDeps.jl:118]: From my understanding,

- `["0"]`: denotes that this dependency has always been used
- `["-0"]`: denotes that after a certain version until the current one this dependency is being used
- `["x-y"]`: denotes during these specific versions a dependency has been used

File: test/runtests.jl (hunk 2 above)
Comment (mattBrzezinski) [reply test/runtests.jl]: Oopsies, that shouldn't even be in there. Just an artifact from making things.

File: src/PkgDeps.jl (hunk 1 above)
Comment (fchorney) [reply src/PkgDeps.jl:118]: ahh ok cool makes sense!

**Comment by mjram0s** [comment]
Should add a test for the case where there are two registries with the same package name

File: test/runtests.jl (hunk 2 above)
Comment (mattBrzezinski) [inline test/runtests.jl]:
```suggestion
using Test
```

**Comment by mattBrzezinski** [comment]
> Should add a test for the case where there are two registries with the same package name

Added in [bb3cae5](https://github.com/mattBrzezinski/PkgDeps.jl/pull/4/commits/bb3cae50838671f06a229409c9ab4c7724bf2380)

File: test/runtests.jl [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    @testset "all registries" begin
+        registries = reachable_registries(; depots=depot)
+
+        @test length(registries) == 2
+    end
+end
+
+@testset "find_upstream_dependencies" begin
+    registry = reachable_registries(; depots=depot)
+    dependents = find_upstream_dependencies("UpDep"; registries=registry)
```
Comment (mjram0s) [inline test/runtests.jl]: I would also add a test for specifying one specific registry e.g. `find_upstream_dependencies("UpDep"; registries="General")` then `@test length(dependents) == 1`

**Review by mjram0s — APPROVED** [review]

**Review by mjram0s — APPROVED** [review]

**Review by fchorney — APPROVED** [review]

**Pull request merged**